### PR TITLE
chore(TS): migrate `Pattern`

### PR DESCRIPTION
**Pull request by ShaMan123**
Migrate Pattern to TS
I thought it had default values in export, that's why I left it but it doesn't so I migrated it.

### CODE REVIEW

File: src/pattern.class.ts [hunk 1]
```diff
@@ -1,196 +1,188 @@
...
-        imageOptions = Object.assign({}, options, { crossOrigin: object.crossOrigin });
-    return fabric.util.loadImage(object.source, imageOptions)
-      .then(function(img) {
-        patternOptions.source = img;
-        return new fabric.Pattern(patternOptions);
-      });
-  };
-})(typeof exports !== 'undefined' ? exports : window);
+  toSVG({ width, height }: TSize) {
+    const patternSource = this.source,
```
Comment (ShaMan123) [inline src/pattern.class.ts:159]: 482ded8b

**Review by ShaMan123 — COMMENTED** [review]
I can't get `.npmignore` to checkout...
Besides that standard es6/TS, except the commented line which is a cleanup for an old PR

File: src/pattern.class.ts [hunk 2]
```diff
@@ -1,196 +1,188 @@
...
+   */
+  isImageSource(): this is TImageSource {
+    return typeof this.source.src === 'string';
+  }
+
+  /**
+   * @returns true if {@link source} is a <canvas> element
+   */
+  isCanvasSource(): this is TCanvasSource {
+    return typeof this.source === 'object' && this.source.toDataURL;
```
Comment (ShaMan123) [inline src/pattern.class.ts:105]: Is this an assertion for a canvas element? Or for a fabric Object? Or both?

File: src/pattern.class.ts [hunk 3]
```diff
@@ -1,196 +1,188 @@
...
+    ].join('\n');
+  }
+  /* _TO_SVG_END_ */
+
+  static async fromObject({ source, ...serialized }: TPatternSerialized, options: TPatternHydrationOptions) {
+    const img = await loadImage(source, {
+      ...options,
+      crossOrigin: serialized.crossOrigin
+    })
+    return new Pattern({ ...serialized, source: img });
```
Comment (asturur) [inline src/pattern.class.ts:183]: everywhere else we use the promise syntax, let's leave that one

**Comment by asturur** [comment]
> Migrate Pattern to TS I thought it had default values in export, that's why I left it but it doesn't so I migrated it.

it could have them :) but patterns are usually small.

**Comment by asturur** [comment]
This is fine for where we are at.
The issue with TS and Classes and that it makes harder to `add whatever i want` to an object ( shadow, fabricObject, pattern gradient ) and export/import it.

Types makes it hard at dev time, classes makes it hard at runtime.

Now Patterns and Gradients are usually not really where people add businnes logic ( names or special props or special behavior ) so those are fine to be standard classes, with or without the default values issue

**Review by asturur — APPROVED** [review]

**Pull request merged**